### ISSUE TITLE
GATv2conv bug in self_loop behaviour in bipartite graphs

### ISSUE THREAD

**Issue by sarihl** [pyg-team/pytorch_geometric#7581]
### 🐛 Describe the bug

When using GATv2conv on a bipartite graph, the self-edges are treated as if they are to be added to the source nodes (which doesn't make sense).
This leads to a fatal error when the target graph is larger (in terms of the number of nodes) than the source graph, where in line 237 of gatv2_conv.py this line appears:
```python
edge_index, edge_attr = add_self_loops(
                    edge_index, edge_attr, fill_value=self.fill_value,
                    num_nodes=num_nodes)
```
where
```python
num_nodes = x_l.size(0)
```
This leads to an error as the target edges have indices that are larger that ```num_nodes``` and therefore this operation fails.

IMHO, I believe that this can be fixed by changing the line in question to:
```python
num_nodes = x_r.size(0)
```

### Environment

* PyG version: 2.3.1
* PyTorch version: 1.13.1
* OS: windows 11
* Python version: 3.7
* CUDA/cuDNN version: 11.6.1
* How you installed PyTorch and PyG (`conda`, `pip`, source): pip
* Any other relevant information (*e.g.*, version of `torch-scatter`): torch-scatter version 2.1.0

**Comment by sarihl** [comment]
I'm on version 2.3.1. But I missed the line you mentioned.
Either way, the code, fails in the case where ```x_r.shape[0] > x_l.shape[0]``` (the number to target nodes is larger the the number of source nodes).

The error occur in ```utils.loop.add_self_loops``` at line 263:
```python
col = edge_index[0] if is_sparse else edge_index[1]  # line 262
loop_attr = scatter(edge_attr, col, 0, N, fill_value)  # line 263
``` 
Since ```col``` uses the indices of the sources of the edges(the larger graph), in graphs as described above, indices will be larger than ```N``` and thus will lead to an error.

Looking at the code, I believe that for Bipartite graphs, the ```add_self_loops``` function should be implemented separately. The current implementation adds the edge indices in lines 244-255 as:
```python
loop_index = torch.arange(0, N, dtype=torch.long, device=edge_index.device)
loop_index = loop_index.unsqueeze(0).repeat(2, 1)
```
This does NOT add self-loops; this adds edges between the i`th node on the left to the i`th node on the right, where IMHO the expected behavior would be to add self-loops to the nodes on the right side of the graph.

If you agree, I would be happy to give this a deeper look and try to implement a fixed version.

**Comment by rusty1s** [comment]
Mh, I agree with your first point. I assume it should be `loop_attr = scatter(edge_attr, col, 0, reduce=fill_value)[:N]` instead.

**Comment by sarihl** [comment]
That would prevent the current error, but if my second point is correct, this solution is incorrect.

I'll try to explain what I think the problem is more clearly. The edge set is extended by adding $[(i,i)]_i$, but a bipartite graph is defined as G(V1+V2, E) such that edges go from V1 to V2 and the input ```edge_index``` has the indices of the sources (inside V1) in ```edge_index[0]``` and the targets (inside V2) in ```edge_index[1]```. Most importantly, each set has its own numbering as of vertices (so vertex number i points to two different nodes depending on whether it is a source or a target).

Hence, when adding the edge $(i,i)$ it is not a self-loop, it is merely an edge connecting two different nodes, one on each side of the graph, which happen to have the same index.

So, if the above is correct, the edges are added incorrectly, and furthermore, the solution you described would be aggregating attributed of edges going out of the source node, and setting that feature as the feature for the "self-loop" for the target node.

note: I believe that the self-loop should be added to the target nodes only, since adding it to the source nodes would not affect the output.

**Comment by rusty1s** [comment]
Yes, that is correct. In case source and destination node describes a different set of nodes, one shouldn't add self loops to the graph at all. That's why we perform additional checks, e.g., in [`HeteroConv`](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/conv/hetero_conv.py#L71-L72).
